README
PerlDAV -- A WebDAV client library for Perl5
    PerlDAV is a Perl library for modifying content on webservers using the
    WebDAV protocol. Now you can LOCK, DELETE and PUT files and much more on
    a DAV-enabled webserver.

    The PerlDAV library consists of:

    *   HTTP::DAV - an object-oriented Web-DAV client API.

    *   dave - the DAV Explorer, an end-user Unix console program for
        interacting with WebDAV servers. dave looks and feels like a
        standard Unix ftp program.

LATEST VERSION AND WHAT'S NEW
    See the included "Changes" file for the full changelog.

INSTALLING PERLDAV
    The lazy way to install PerlDAV:

       $ perl -MCPAN -e shell
       cpan> install HTTP::DAV

    Or the normal way:

    Retrieve the latest copy from CPAN:
    http://search.cpan.org/dist/HTTP-DAV/

       $ perl Makefile.PL # Creates the Makefile
       $ make             # Runs the makefile
       $ make test        # Optional (See Interopability below)
       $ make install     # Installs dave and HTTP::DAV

    With this method you will first have to install the pre-requisites: LWP
    and XML::DOM, see "what are the prerequisites?".

    When you install PerlDAV, the HTTP::DAV library will be installed to
    your Perl library location (usually /usr/local/lib/perl5)

    "dave" will be installed to /usr/local/bin. This suits most people but
    you can modify this by using the INSTALLBIN flag:

       $ perl Makefile.PL INSTALLBIN="/home/pcollins/bin"

  What Are The Prerequisites?
    *   Perl 5.6.0+

    *   LWP (Have not tested lower than v5.48)

    *   Scalar::Util (standard library from 5.8.0+)

    *   XML::DOM (Have not tested lower than v1.26). Requires James Clark's
        expat library:

    *   To access SSL urls you will need Crypt::SSLeay and/or
        IO::Socket::SSL.

  Optional Prerequisites.
    *   Crypt::SSLeay if you'd like to use https. Crypt::SSLeay requires the
        openssl library as well. See Crypt::SSLeay's excellent install
        instructions for how to get https support into LWP (and hence
        HTTP::DAV). I've tested HTTP::DAV and Crypt::SSLeay against
        Apache/mod_dav with the mod_ssl plugin. Works seamlessly.

    *   MD5 if you'd like to use LWP's Digest authentication.

    To get the latest versions of these prerequisite modules you can simply
    type this at the command prompt:

       $ <install expat>
    then:
       $ perl -MCPAN -e shell
       cpan> install LWP
       cpan> install XML::DOM

    or if you just 'install HTTP::DAV' the lovely CPAN module should just
    magically install all of the prerequisites for you (you'll still need to
    manually instal expat though).

  What Systems Does It Work With?
    HTTP::DAV and dave are pure perl so only needs Perl 5.6.0 (or later).
    PerlDAV is known to run under Windows (although I haven't tried it
    myself) and should run under all Unix systems.

WHERE ARE THE MANUALS?
    Once you've installed PerlDAV, you can type:

       $ perldoc HTTP::DAV
       $ man dave

GETTING HELP
  The perldav mailing list
    There is a mailing list for PerlDAV for use by Developers and Users.
    Please see http://mailman.webdav.org/mailman/listinfo/perldav

THE TEST SUITE & INTEROPERABILITY
    You will notice that the standard "make test" command invokes a large
    set of test procedures, but most will be skipped. This standard test is
    sufficient to give you a good indication that PerlDAV has installed
    properly.

    If you'd like to see how well PerlDAV performs against a particular DAV
    server then you should set the URL (and username,password) in the test
    suite t/TestDetails.pm. Then you can run "make test" again and watch the
    test suite perform as many operations as the server supports. Please
    note that the test suite will perofrm well over 200 HTTP requests to
    your server.

    I have tested PerlDAV against IIS5, mod_dav and the Xythos WFS.

    Out of the box, the test suite should NOT fail on any tests.

    The test suite is the best way to test interopability between PerlDAV
    and other servers. I'd really like help with testing PerlDAV's
    interoperability. So if one or more tests fail against your server
    please follow the following steps:

    *   Determine which test is failing.

    *   set DEBUG to on: edit the script and change HTTP::DAV::DebugLevel(0)
        to (3).

    *   Delete previous server output: rm /tmp/perldav_debug.txt

    *   Run that single test again:

        $make test TEST_FILES=t/thetest.t TEST_VERBOSE=1 > testoutput.log

    *   Then gzip and mail me both testoutput.log and /tmp/perldav_debug.txt
        with details of the test environment. (My email is at the bottom)

    Alternatively, you could have a shot at solving the bug yourself :)

BUGS and TODO
    Need to convert XML::DOM to a DOM Level 2 compliant parser like
    XML::Gdome.

See TODO for what is left to be done.
AUTHOR AND COPYRIGHT
    This module is Copyright (C) 2001 by

       Patrick Collins
       G03 Gloucester Place, Kensington
       Sydney, Australia

       mailto:pcollins@cpan.org
       Phone: +61 2 9663 4916

    All rights reserved.

MAINTAINER
    The current maintainer of HTTP-DAV is Cosimo Streppone for Opera
    Software ASA. You can contact us at "cosimo@cpan.org" or
    "opera@cpan.org".

LICENSE
    You may distribute this module under the terms of either the GNU General
    Public License or the Artistic License, as specified in the Perl README
    file.

    $Id$

** This file was automatically generated from **
** doc/Changes.pod. To edit it, see there.    **


Changes
Revision history for HTTP::DAV
  v0.49 (released 2018/11/28):
    *   bug fixes

        Fixed perl shebang line in "dave" script, for ExtUtils::MakeMaker to
        correctly replace it. Closes "RT#127819".

        Fixed "propfind" response handling to also consider successful an
        HTTP 207 status code. Closes "RT#127591".

  v0.48 (released 2015/03/26):
    *   bug fixes

        "RT#103126", fixed faulty code to add trailing slash to URLs.

  v0.47 (released 2012/03/24):
    *   bug fixes

        Improve "propfind()" resilience when server response contains
        broken, truncated or no XML at all. RT#75011.

  v0.46 (released 2012/01/11):
    *   improvements

        HTTP::DAV should now be working with more WebDAV servers.

        We are more flexible in what content types we consider to be XML.
        Thanks Ron1 and Adam for the feedback and patches.

  v0.45 (released 2011/09/18):
    *   bug fixes

          - Fixed RT #69439 (http://rt.cpan.org/Public/Bug/Display.html?id=69439),
            insecure /tmp files handling in dave client.

    *   improvements

          - Added -tmpdir option to dave client.
          - Reorganized distribution layout to match usual CPAN practice
          - Removed remains of svn-era ($Id and such...)

  v0.44 (released 2011/06/19):
    *   bug fixes

          - Fixed RT #68936 (http://rt.cpan.org/Public/Bug/Display.html?id=68936),
            Fixed errors() method that would bomb out when the "_errors" attribute
            wasn't initialized. Thanks to Michael Lackoff for reporting.

  v0.43 (released 2011/04/12):
    *   bug fixes

          - Fixed RT #38677 (http://rt.cpan.org/Public/Bug/Display.html?id=38677),
            Intercept correctly 405 (Method now allowed) errors and report them
            to the clients.

  v0.42 (released 2010/11/07):
    *   bug fixes

          - Fixed RT #60457 (http://rt.cpan.org/Public/Bug/Display.html?id=60457),
            Added and documented possibility to pass your own custom HTTP headers.
          - Fixed errors in the code examples in the synopsis.

  v0.41 (released 2010/07/24):
    *   bug fixes

          - Fixed RT #59674 (http://rt.cpan.org/Public/Bug/Display.html?id=59674),
                When SSL support is needed but not installed, a more specific
            error messages is now displayed, instead of "not DAV enabled or not accessible".

  v0.40 (released 2010/01/27):
    *   bug fixes

          - Fixed RT #47500 (http://rt.cpan.org/Public/Bug/Display.html?id=47500),
                HTTP::DAV::Comms->credentials() method erroneously autovivified
                basic authentication internal values, causing wrong or undefined
                credentials to be sent out, or credentials to be "forgot" by HTTP::DAV.

  v0.39 (released 2009/12/12):
    *   bug fixes

          - Fixed RT #52665 (http://rt.cpan.org/Public/Bug/Display.html?id=52665),
                Using dave or propfind() on URLs containing escaped chars (%xx) could fail,
                due to upper/lower case differences. Thanks to cebjyre for the patch
                and the test case.

  v0.38 (released 2009/06/09):
    *   bug fixes

          - Fixed RT #14506 (http://rt.cpan.org/Public/Bug/Display.html?id=14506),
            about the missing get_lastresponse() method. It was a documentation bug.

          - Fixed RT #29788 (http://rt.cpan.org/Public/Bug/Display.html?id=29788),
            avoid file corruptions on Win32 when calling HTTP::DAV::get() method.

          - Fixed RT #31014 (http://rt.cpan.org/Public/Bug/Display.html?id=31014),
            probably already in v0.34, since it seems related to propfind() "depth" bug.

  v0.37 (released 2009/03/24):
    *   bug fixes

          - Fixed RT #44409 (http://rt.cpan.org/Public/Bug/Display.html?id=44409),
            Small bug in HTTP::DAV::put(). Passing a reference as local content resulted
                in the "SCALAR(0x12345678)" being logged instead of the real scalar.

  v0.36 (released 2009/02/25):
    *   bug fixes

          - Fixed RT #19616 (http://rt.cpan.org/Public/Bug/Display.html?id=19616),
            LWP::UserAgent::redirect_ok() is not changed anymore. We're subclassing
            it from HTTP::DAV::UserAgent and overriding redirect_ok() there.

          - Fixed RT #42877 (http://rt.cpan.org/Public/Bug/Display.html?id=42877),
            HTTP::DAV::UserAgent::credentials() has been modified to behave like
            LWP::UserAgent::credentials(), otherwise basic authentication breakages
            can occur.

          - Fixed a problem with C<-depth> argument to C<HTTP::DAV::propfind()> that
            could lead to massive performance degradation, especially when running
                C<propfind()> against large folders.
                C<-depth> was set to 1 even when passed as zero.

  v0.35 (released 2008/11/03):
    *   bug fixes

          - Fixed RT #40318 (http://rt.cpan.org/Public/Bug/Display.html?id=40318),
            about getting single or multiple files directly to \*STDOUT.

  v0.34 (released 2008/09/11):
    *   bug fixes

          - Fixed RT #39150 (http://rt.cpan.org/Public/Bug/Display.html?id=39150),
            about downloading multiple files in the same directory.

  v0.33 (released 2008/08/24):
    *   documentation

          - Clearly state that opera software asa is now co-maintainer of http::dav

          - Fixed various inconsistencies in the v0.32 documentation

  v0.32 (released 2008/08/24):
    *   incompatibilities

         - Now HTTP::DAV requires Perl 5.6.0+ and Scalar::Util (core in 5.8.x).

    *   bug fixes

         - Now HTTP::DAV objects are correctly released from memory when
           they go out of scope. Now it should be possible to use multiple instances
           of HTTP::DAV even in long-running processes.

           Was caused by circular references between HTTP::DAV and HTTP::DAV::Resource.

  v0.31 (released 2002/04/13):
    *   Apache 2 mod_dav support


DAV/Response.pm
# $Id: Response.pm,v 0.13 2002/04/06 17:45:42 pcollins Exp $
package HTTP::DAV::Response;

$VERSION = sprintf("%d.%02d", q$Revision: 0.13 $ =~ /(\d+)\.(\d+)/);

use strict;
use vars qw(@ISA);
use vars qw($VERSION);

require HTTP::Response;
@ISA = qw( HTTP::Response );

my %dav_status_codes = (
   102 => "Processing. Server has accepted the request, but has not yet completed it",
   204 => "No Content",
   207 => "Multistatus",
   422 => "Unprocessable Entity. Bad client XML sent?",
   423 => "Locked. The source or destination resource is locked",
   424 => "Failed Dependency",
   507 => "Insufficient Storage. The server is unable to store the request",
);

# PROTECTED METHODS

sub clone_http_resp {
   my ($class,$http_resp) = @_;
   my %clone = %{$http_resp};
   my $self = \%clone;
   bless $self, (ref($class) || $class); 
}

# This routine resets the base
# message in the 
# object based on the 
# code and the status_codes above.
# set_message('207');
sub set_message {
   my ($self,$code) = @_;

   # Set the status code
   if ( defined $dav_status_codes{$code} ) {
      $self->message( $dav_status_codes{$code} );
   }
}

sub set_responsedescription {
   $_[0]->{'_dav_responsedescription'} = $_[1] if $_[1];
}

sub get_responsedescription { $_[0]->{'_dav_responsedescription'}; }

sub add_status_line {
   my($self,$message,$responsedescription,$handle,$url) = @_;
   
   # Parse "status-line". See section 6.1 of RFC 2068
   # Status-Line= HTTP-Version SP Status-Code SP Reason-Phrase CRLF
   if (defined $message && $message =~ /^(.*?)\s(.*?)\s(.*?)$/ ) {
      my ($http_version,$status_code,$reason_phrase) = ($1,$2,$3);

      push ( @{$self->{_dav_multistatus}},
               {
                  'handle' => $handle,
                  'url' => $url,
                  'HTTP Version' => $http_version,
                  'code' => $status_code,
                  'message' => $reason_phrase,
                  'description' => $responsedescription,
               }
           );
      return 1;

   } else {
      return 0;
   }
}

# PUBLIC METHODS

sub is_multistatus {
    return ($_[0]->code eq "207" )? 1:0; 
}

sub messages {
   my ($self) = @_;

   my @messages = ();
   if ($self->is_multistatus() ) {
      foreach my $num ( 0 .. $self->response_count()) {
         push(@messages, $self->message_bynum($num));
      }
   } else {
      push(@messages,$self->message());
   }

   return wantarray ? @messages : join("\n",@messages);
}

sub codes {
   my ($self) = @_;

   my @codes = ();
   if ($self->is_multistatus() ) {
      foreach my $num ( 0 .. $self->response_count()) {
         push(@codes, $self->code_bynum($num));
      }
   } else {
      push(@codes,$self->code());
   }

   return \@codes;
}

sub response_count { 
   return -1 unless exists $_[0]->{_dav_multistatus};
   return -1 unless ref($_[0]->{_dav_multistatus}) =~ /ARRAY/;
   return  $#{$_[0]->{_dav_multistatus}}; 
}

sub message_bynum { $_[0]->{_dav_multistatus}[$_[1]]{'message'}; }
sub    code_bynum { $_[0]->{_dav_multistatus}[$_[1]]{'code'}; }
sub     url_bynum { $_[0]->{_dav_multistatus}[$_[1]]{'url'}; }
sub description_bynum { $_[0]->{_dav_multistatus}[$_[1]]{'description'}; }

sub response_bynum {
   my ($self,$number) = @_;

   if (defined $number && $number>=0 ) {
      return (
         $self->code_bynum($number),
         $self->message_bynum($number),
         $self->url_bynum($number),
         $self->description_bynum($number),
      );
   }
}

sub is_success {
   my ($self) = @_;

   if ($self->is_multistatus() ) {
      foreach my $code ( @{ $self->codes() } ) {
         return 0 if ( HTTP::Status::is_error($code) );
      }
   } else {
      return ($self->SUPER::is_success() || 0);
   }

   return 1;
}

sub as_string {
   my ($self) = @_;
   my ($ms, $returnstr) = "";

#   use Data::Dumper;
#   print Data::Dumper->Dump( [\$self] , [ '$self' ] );
   foreach my $num ( 0 .. $self->response_count() ) {
      my %h = %{$self->{_dav_multistatus}[$num]};
      $ms .= "Error number $num ($h{handle}):\n";
      $ms .= "   Href:       $h{url}\n"                 if defined $h{url};
      $ms .= "   Mesg(code): $h{message} ($h{code})\n"  if defined $h{code};
      $ms .= "   Desc:       $h{'description'}\n"       if defined $h{'description'};
      $ms .= "\n";
   }

   my $rd = $self->get_responsedescription() || "";

   $returnstr .= "Multistatus lines:\n$ms\n" if $ms;
   $returnstr .= "Overall responsedescription: \"$rd\"\n" if $rd;
   $returnstr .= $self->SUPER::as_string;
   $returnstr;
}

=head1 NAME

HTTP::DAV::Response - represents a WebDAV HTTP Response (ala HTTP::Response)

=head1 SYNOPSIS

require HTTP::DAV::Response;

=head1 DESCRIPTION

The HTTP::DAV::Response class encapsulates HTTP style responses.  A response consists of a response line, some headers, and (potentially empty) content. 

HTTP::DAV::Response is a subclass of C<HTTP::Response> and therefore inherits its methods.  (HTTP::Response in turn inherits it's methods from C<HTTP::Message>).

Therefore, this class actually inherits a rich library of functions. You are more likely wanting to read the C<HTTP::Response> class as opposed to this class.

Instances of this class are usually created by a C<HTTP::DAV::Resource> object after it has performed some request (such as get, lock, delete, etc). You use the object to analyse the success or otherwise of the request.

HTTP::DAV::Response was created to handle two extra functions that normal HTTP Responses don't require:

 - WebDAV reponses have 6 extra error codes: 102, 207, 422, 423, 424 and 507. Older versions of the LWP's C<HTTP::Status> class did not have these extra codes. These were added.

 - WebDAV responses can actually contain more than one response (and often DO contain more than one) in the form of a "Multistatus". These multistatus responses come in the form of an XML document. HTTP::DAV::Response can accurately parse these XML responses and emulate the normal of the C<HTTP::Response>.

HTTP::DAV::Response transparently implements these extra features without the user having to be aware, so you really should be reading the C<HTTP::Response> documentation for most of the things you want to do (have I already said that?).

There are only a handful of custom functions that HTTP::DAV::Response returns and those are to handle multistatus requests, C<messages()> and C<codes()>.

The six extra status codes that DAV servers can be returned in an HTTP Response are:
  102 => "Processing. Server has accepted the request, but has not yet completed it",
  207 => "Multistatus",
  422 => "Unprocessable Entity. Bad client XML sent?",
  423 => "Locked. The source or destination resource is locked",
  424 => "Failed Dependency",
  507 => "Insufficient Storage. The server is unable to store the request",

See C<HTTP::Status> for the rest.

=head1 HANDLING A MULTISTATUS 

So, many DAV requests may return a multistatus ("207 multistatus") instead of, say, "200 OK" or "403 Forbidden".

The HTTP::DAV::Response object stores each "response" sent back in the multistatus. You access them by array number.

The following code snippet shows what you will normally want to do:

...
$response = $resource->lock();

if ( $response->is_multistatus() ) {

   foreach $num ( 0 .. $response->response_count() ) {
      ($err_code,$mesg,$url,$desc) = 
         $response->response_bynum($num);
      print "$mesg ($err_code) for $url\n";
   }
}

Would produce something like this:
   Failed Dependency (424) for /test/directory
   Locked (423) for /test/directory/file3

This says that we couldn't lock /test/directory 
because file3 which exists inside is already locked by somebody else.

=head1 METHODS

=over 4 

=item B<is_multistatus>

This function takes no arguments and returns a 1 or a 0.

For example: if ($response->is_multistatus() ) { }

If the HTTP reply had "207 Multistatus" in the header then that indicates that there are multiple status messages in the XML content that was returned.

In this event, you may be interested in knowing what the individual messages were. To do this you would then use C<messages>. 

=item B<response_count>

Takes no arguments and returns "the number of error responses -1" that we got.
Why -1? Because usually you will want to use this like an array operator:

foreach $num ( 0 .. $response->response_count() ) { 
   print $response->message_bynum();
}

=item B<response_bynum>

Takes one argument, the "response number" that you're interested in. And returns an array of details:

   ($code,$message,$url,$description) = response_bynum(2);

where 
   $code - is the HTTP error code (e.g. 403, 423, etc).
   $message - is the associated message for that error code.
   $url - is the url that this error applies to (recall that there can be multiple responses within one response and they all relate to one URL)
   $description - is server's attempt at an english description of what happened.

=item B<code_bynum>

Takes one argument, the "response number" that you're interested in, and returns it's code. E.g:

  $code = $response->code_bynum(1);

See C<response_bynum()>

=item B<message_bynum>

Takes one argument, the "response number" that you're interested in, and returns it's message. E.g:

  $code = $response->message_bynum(1);

See C<response_bynum()>

=item B<url_bynum>

Takes one argument, the "response number" that you're interested in, and returns it's url. E.g:

  $code = $response->message_bynum(1);

See C<response_bynum()>

=item B<description_bynum>

Takes one argument, the "response number" that you're interested in, and returns it's description. E.g:

  $code = $response->message_description(1);

See C<response_bynum()>

=item B<messages>

Takes no arguments and returns all of the messages returned in a multistatus response. If called in a scalar context then all of the messages will be returned joined together by newlines. If called in an array context the messages will be returned as an array.

$messages = $response->messages();
e.g. $messages eq "Forbidden\nLocked";

@messages = $response->messages();
e.g. @messages eq ["Forbidden", "Locked"];

This routine is a variant on the standard C<HTTP::Response> C<message()>. 

=cut

